raster/src/main/scala/geotrellis/raster/CellFeatures.scala

```diff
@@ -0,0 +1,126 @@
+/*
+ * Copyright 2019 Azavea
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package geotrellis.raster
+
+import geotrellis.raster.rasterize.Rasterizer
+import geotrellis.vector.{Feature, Geometry, Point, Polygon, Extent}
+import geotrellis.util.MethodExtensions
+
+import spire.syntax.cfor._
+
+
+/** Type class to convert a raster into features of cell geometries */
+trait CellFeatures[R, D] {
+
+  /** Describe cell grid of the source raster */
+  def cellGrid(raster: R): GridExtent[Long]
+
+  /** Given a geometry, produce an iterator of cell features intersecting this geometry.
+    * Relating geometry to raster cell grid is rasterization, options for this are given.
+    * The call will decide how each cell will be represented as geometry (Point, Polygon)
+    *
+    * @note This method returns an Iterator because the feature representation of raster will require significantly
+    *   more memory and it may not be possible to hold all features from a "large" raster tile in memory.
+    *   It is assumed that either filter, fold or sink will be the next step.
+    *
+    * @note Conceptually rasterisation must happen but it is left to interface implementer to decide how to do it.
+    *   Potentially the raster will need to be read in chunks.
+    *
+    * @param raster   source of cell values
+    * @param geom     geometry filter, only cells under this geometry will be returned
+    * @param options  rasterizer options that specify which cells should be considered as intersection geom
+    * @param cellGeom function that converts a cell, specified by column and row into geometry (Point, Polygon)
+    */
+  def cellFeatures[G <: Geometry](raster: R, geom: Geometry, options: Rasterizer.Options, cellGeom: (Long, Long) => G): Iterator[Feature[G, D]]
+}
+
+object CellFeatures {
+  def apply[R, D](implicit ev: CellFeatures[R, D]): CellFeatures[R, D] = ev
+
+  /** Produce a CellFeatures instance given functions to retrieve a values and describe the cell grid.
+    * This function is suitable for use with rasters that fit entirely in memory.
+    */
+  def make[R, D](getGrid: R => GridExtent[Int], cellValue: (R, Int, Int) => D): CellFeatures[R, D] =
```

Comment:
`cellValue` is `Function3` and is not specialized, causing boxing for col/row here. Elsewhere in the code we would make a specific trait. However since the main purpose of this interface is to turn a single double cell value into a nested class with a whole geometry ... its fine, we'll let the boxing slide here.

Reply:
It is itself a slow operation, would like to see someday how bad is boxing here.